docs: plugin authoring workflow overview pages (#20847)
gatsbyjs/gatsby: docs/docs/creating-a-local-plugin.md

If a plugin is only relevant to your specific use-case, or if you’re developing a plugin and want a simpler workflow, a locally defined plugin is a convenient way to create and manage your plugin code.

## Project structure for a local plugin

Place the code in the `plugins` folder in the root of your project like this:

```
/my-gatsby-site
└── gatsby-config.js
└── /src
└── /plugins
    └── /my-own-plugin
        └── package.json
```

The plugin also needs to be added to your `gatsby-config.js`, because there is no auto-detection of plugins. It can be added alongside any other 3rd party Gatsby plugins already included in your config.

> For the plugin to be discovered when you run `gatsby develop`, the plugin's root folder name needs to match the name used in the `gatsby-config.js` (_not_ its _name_ in its `package.json` file). For example, in the above structure, the correct way to load the plugin is:

Reviewer: ```suggestion
For the plugin to be discovered when you run `gatsby develop`, the plugin's root folder name needs to match the name used in the `gatsby-config.js` (_not_ the _name_ it goes by in your `package.json` file). For example, in the above structure, the correct way to load the plugin is:
```

```javascript:title=gatsby-config.js
module.exports = {
  plugins: [
    `gatsby-third-party-plugin`,
    `my-own-plugin`, // highlight-line
  ],
}
```

> Then your plugin can begin to hook into Gatsby through Node and SSR APIs.

Reviewer: ```suggestion
Then the plugin can begin to hook into Gatsby through Node and SSR APIs.
```
Not sold on this one. Seems like it could go either way?

> Then the plugin can begin to hook into Gatsby through Node and SSR APIs.

Reviewer: ```suggestion
Then the plugin can begin to hook into Gatsby through [Node](/docs/node-apis/) and [SSR](/docs/ssr-apis/) APIs.
```

## Developing a local plugin that is outside your project

> Your plugin doesn't have to be in your project in order to be tested or worked on. If you'd like to [decouple](/docs/glossary#decoupled) your plugin from your site so it can be published as its own package, or you'd like to test or develop a forked version of a community authored plugin, you can follow one of the methods described below.

Reviewer: ```suggestion
Your plugin doesn't have to be in your project in order to be tested or worked on. If you'd like to [decouple](/docs/glossary#decoupled) your plugin from your site you can follow one of the methods described below. This is a useful thing to do if you want to publish the plugin as its own package, or test/develop a forked version of a community authored plugin.
```

### Using `require.resolve` and a filepath

> In addition to including a plugin in the `plugins` folder, you can give the path to a plugin included in a different location on your machine and `require` it.

Reviewer: ```suggestion
Including a `plugins` folder is not the only way to reference a local plugin. Alternatively, you can include a plugin in your `gatsby-config.js` file by directly referencing its relative path and then `require` it.
```
Just an idea? I think a big question here will be what is the path relative to. Is it relative to the config file?

Author: Good point! I like how you wrote this, I'll add a little mention of how it's relative to the config.

```javascript:title=gatsby-config.js
module.exports = {
  plugins: [
    `gatsby-plugin-react-helmet`,
    // highlight-start
    {
      // including a plugin from outside the plugins folder needs the path to it
      resolve: require.resolve(`../path/to/gatsby-local-plugin`),
    },
    // highlight-end
  ],
}
```

### Using `npm link` or `yarn link`

> You can use [`npm link`](https://docs.npmjs.com/cli/link.html) or [`yarn link`](https://yarnpkg.com/lang/en/docs/cli/link/) to reference a package from another location on your machine in another project.

Reviewer: What do you mean by "another project" here? Do you mean the package is in another project? Or is the Gatsby site your other project in this case?

Author: Yeah this does sound a little ambiguous/redundant.  I think it could be reduced to this:

```suggestion
You can use [`npm link`](https://docs.npmjs.com/cli/link.html) or [`yarn link`](https://yarnpkg.com/lang/en/docs/cli/link/) to reference a package from another location on your machine.
```

I think it's kind of saying the same thing twice, how's that suggestion?

Reviewer: Looks good to me!

> By running `npm link ../path/to/my-plugin` in the root of your Gatsby site where you would like to use the plugin, your computer will create a symlink to your package.

Reviewer: ```suggestion
By running `npm link ../path/to/my-plugin` in the root of your Gatsby site, your computer will create a symlink to your package.
```
I think we can assume we're referencing the site that would make use of the plugin. Allows this sentence to be a bit clearer.

This is a similar process to setting up yarn workspaces for development with Gatsby themes (which is the recommended approach for developing themes). You can read how to setup a site in this manner in the [Building a Theme guide](/tutorial/building-a-theme/#set-up-yarn-workspaces).

> **Note**: you can see an example for using a local plugin from the plugins folder, with `require.resolve`, and `npm link` demonstrated in [this example repository](https://github.com/gillkyle/local-plugins-in-gatsby).

Reviewer: ```suggestion
**Note**: See an example of using a local plugin from the plugins folder, with `require.resolve`, and `npm link` in [this example repository](https://github.com/gillkyle/local-plugins-in-gatsby).
```
